pace-neutrons/horace-euphonic-interface
# Update python wrapper for help functions

Description
---------------

Updates the code to v0.2.1 of [light_python_wrapper](https://github.com/pace-neutrons/light_python_wrapper/tree/v0.2.1). This now contains overloaded versions of the `help` and `doc` functions which will display the Python docstrings in Matlab.

Because Matlab puts the path of any toolboxes at the *end* of the path list, these functions will always be shadowed by the built-in `help` / `doc` functions for installed toolboxes. Hence for the horace-euphonic-interface, we put these functions in the `+euphonic` package namespace. If users then specifically import them using either `import euphonic.help` or `import euphonic.*` they will take [precedence](https://www.mathworks.com/help/matlab/matlab_prog/function-precedence-order.html) over the built-in functions. The web documentation has been updated to note this.

The overloaded `help` methods of the wrapped classes still work but they require an actual object instance. To make this work for the static `from_*` methods, these are now declared as constant *properties* (not methods - because in Matlab, specifying any function means calling it; only for properties [e.g. scalar variables] can we just specify the object without calling). So, the `from_` methods now behave in a "Pythonic" fashion: `euphonic.ForceConstants.from_phonopy` returns a wrapped reference to the Python `from_phonopy` method, whilst `euphonic.ForceConstants.from_phonopy()` [with explicit brackets] will call the Python method without arguments. Previously when `from_phonopy` was defined as a Matlab static method, `euphonic.ForceConstants.from_phonopy` will call the Matlab function with no arguments (which then called the Python method with no arguments - in this specific case this is a valid call so we still want this overload; for the other `from_` methods calling it without arguments was invalid).

To test
---------

Check that automatic builds pass test
Install the `mltbx` toolbox created and check that it works
Try:

```
help(euphonic.ForceConstants.from_castep)
euphonic.help euphonic.ForceConstants

import euphonic.*
help euphonic.ForceConstants.from_castep
doc euphonic
doc euphonic.ForceConstants
```

And check that the internal hyperlinks work (this will only work if the `import` is used; e.g. `euphonic.help euphonic.ForceConstants` will generate bad links). 

Check that calling help on the class (e.g. `help(euphonic.CoherentCrystal)`) will also show the docstring of the `__init__` method.
